contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md

```diff
@@ -0,0 +1,285 @@
+view of the cluster.
+
+#### Configurator
+
+A [configurator](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/factory.go#L90)
+builds the scheduler instance by wiring plugins, cache, queues,
+handlers and other elements together. Each profile [is initialized](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/factory.go#L138-L147)
+with its own framework (with all frameworks sharing informers, event recorders, etc.).
+
+At this point it’s still possible to have the configurator create the instance
+[from a policy file](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/factory.go#L213).
+Though, this approach is deprecated and will be removed
+from the configuration eventually. Keeping only the kube scheduler configuration
+as the only way to provide the configuration.
+
+#### Default scheduling algorithm
+
+The codebase defines a [ScheduleAlgorithm](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L61-L66)
+interface.
+Any implementation of the interface can be used as a scheduling algorithm.
+There are two methods:
+- `Schedule`: responsible for scheduling a pod using plugins from `PreFilter`
+  up to `NormalizeScore` extension points, provides [ScheduleResult](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L70-L77)
+  containing a scheduling decision (the most suitable nodes) with additional
+  accompanying information such as how many nodes were evaluated
+  and how many nodes were found feasible for scheduling.
+- `Extenders`: currently exposed only for testing
+
+Each cycle of the default algorithm implementation consists of:
+1. Taking the [current snapshot](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L101)
+   from the scheduling cache
+1. [Filter out all nodes](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L110)
+   not feasible for scheduling a pod
+   1. Run [PreFilter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L230)
+      first (preprocessing phase, e.g. computing pod [anti-]affinity relations)
+   1. Run [Filter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/core/generic_scheduler.go#L261) in parallel:
+      filter out all nodes which does not satisfy pod’s constraints
+      (e.g. sufficient resources, node affinity, etc.), including running filter extenders.
+   1. Run [PostFilter plugins](https://github.com/kubernetes/kubernetes/blob/a651804427dd9a15bb91e1c4fb7a79994e4817a2/pkg/scheduler/scheduler.go#L479)
```

Comment:
Run PostFilter plugins if no node can fit the incoming pod.

Reply:
Updated